src/os/spawn.php

```diff
@@ -0,0 +1,173 @@
+<?hh
+
+namespace HH\Lib\Experimental\OS;
+use namespace HH\Lib\C;
+use namespace HH\Lib\Vec;
+use namespace HH\Lib\_Private\_OS;
+use type HH\Lib\OS\{FileDescriptor, IsNotADirectoryException};
+use function HH\Lib\OS\{close, open};
+
+newtype pid_t as int = int;
+newtype ExitCode as int = int;
+
+<<__Sealed(
+  posix_spawn_file_actions_addopen::class,
+  posix_spawn_file_actions_adddup2::class,
+  posix_spawn_file_actions_addchdir_np::class,
+)>>
+interface PosixSpawnFileActionsSetter {}
+
+final class posix_spawn_file_actions_addopen
+  implements PosixSpawnFileActionsSetter {
+  public function __construct(
+    public int $filedes,
+    public string $path,
+    public int $oflag,
+    public int $mode,
+  ) {}
+}
+
+final class posix_spawn_file_actions_adddup2
+  implements PosixSpawnFileActionsSetter {
+  public function __construct(
+    public FileDescriptor $filedes,
+    public int $newfiledes,
+  ) {}
+}
+
+final class posix_spawn_file_actions_addchdir_np
+  implements PosixSpawnFileActionsSetter {
+  public function __construct(public string $path) {
+  }
+}
+
+type posix_spawn_file_actions_t = vec<PosixSpawnFileActionsSetter>;
+
+const int POSIX_SPAWN_SETPGROUP = 2;
+const int POSIX_SPAWN_SETSID = 128;
+
+type PosixSpawnFlags = int;
+
+type posix_spawnattr_t = shape(
+  ?'posix_spawnattr_setpgroup' => pid_t,
+  ?'posix_spawnattr_setflags' => PosixSpawnFlags,
+);
+
+function posix_spawnp(
+  string $file,
+  posix_spawn_file_actions_t $file_actions,
+  posix_spawnattr_t $attributes,
+  vec<string> $argv,
+  vec<string> $envp,
+): pid_t {
+
+  // $default_fork_and_execve will call _OS\fork_and_execve but redirect
+  // standard I/O to /dev/null by default.
+  $default_fork_and_execve = (
```

Comment:
this is a very long lambda definition; split it out to a helper function in `_OS\`